### PR TITLE
ovl_En_Brob OK

### PR DESCRIPTION
**Pull request by Emill**

### CODE REVIEW

File: src/overlays/actors/ovl_En_Brob/z_en_brob.c [hunk 1]
```diff
@@ -54,37 +61,285 @@ static ColliderCylinderInit D_809CBA80 = {
     },
     { 8000, 11000, -5000, { 0, 0, 0 } },
 };
-*/
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Brob/EnBrob_Init.s")
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Brob/EnBrob_Destroy.s")
+static CollisionCheckInfoInit sCylinderInit = { 0, 60, 120, MASS_IMMOVABLE };
```
Comment (petrie911) [inline src/overlays/actors/ovl_En_Brob/z_en_brob.c:65]: This should be named `sColChkInfoInit`. `sCylinderInit` should be the collider init.

File: src/overlays/actors/ovl_En_Brob/z_en_brob.c [hunk 2]
```diff
@@ -54,37 +61,285 @@ static ColliderCylinderInit D_809CBA80 = {
...
+    Collider_DestroyCylinder(globalCtx, &this->colliders[1]);
+}
+
+void func_809CADDC(EnBrob* this, GlobalContext* globalCtx) {
+    func_8003EC50(globalCtx, &globalCtx->colCtx.dyna, this->dyna.bgId);
+    if (this->updateFunc == func_809CB2B8) {
+        this->timer = 200;
+    } else {
+        this->timer = 0;
+    }
```
Comment (petrie911) [inline src/overlays/actors/ovl_En_Brob/z_en_brob.c:122]: possible ternary?

File: src/overlays/actors/ovl_En_Brob/z_en_brob.c [hunk 3]
```diff
@@ -54,37 +61,285 @@ static ColliderCylinderInit D_809CBA80 = {
...
+void func_809CAEF4(EnBrob* this) {
+    Animation_MorphToPlayOnce(&this->skelAnime, &D_06000290, -5.0f);
+    this->unk_1AE -= 125.0f;
+    Actor_SetColorFilter(&this->dyna.actor, 0, 0xFF, 0, 0x50);
+    Audio_PlayActorSound2(&this->dyna.actor, NA_SE_EN_GOMA_JR_FREEZE);
+    this->updateFunc = func_809CB2B8;
+}
+
+void func_809CAF88(EnBrob* this) {
+    Animation_Change(&this->skelAnime, &D_06001750, -1.0f, Animation_GetLastFrame(&D_06001750), 0.0f, 2, -5.0f);
```
Comment (petrie911) [inline src/overlays/actors/ovl_En_Brob/z_en_brob.c:150]:
```suggestion
    Animation_Change(&this->skelAnime, &D_06001750, -1.0f, Animation_GetLastFrame(&D_06001750), 0.0f, ANIMMODE_ONCE, -5.0f);
```

File: src/overlays/actors/ovl_En_Brob/z_en_brob.c [hunk 4]
```diff
@@ -54,37 +61,285 @@ static ColliderCylinderInit D_809CBA80 = {
...
+        }
+    } else if (this->timer >= 81) {
+        this->dyna.actor.colorFilterTimer = 80;
+    }
+}
+
+void func_809CB114(EnBrob* this, GlobalContext* globalCtx) {
+    f32 curFrame;
+
+    if (SkelAnime_Update(&this->skelAnime) != 0) {
```
Comment (petrie911) [inline src/overlays/actors/ovl_En_Brob/z_en_brob.c:180]:
```suggestion
    if (SkelAnime_Update(&this->skelAnime)) {
```

File: src/overlays/actors/ovl_En_Brob/z_en_brob.c [hunk 5]
```diff
@@ -54,37 +61,285 @@ static ColliderCylinderInit D_809CBA80 = {
...
+        func_809CAEA0(this);
+        return;
```
Comment (petrie911) [inline src/overlays/actors/ovl_En_Brob/z_en_brob.c:182]: use an `else` here.

File: src/overlays/actors/ovl_En_Brob/z_en_brob.c [hunk 6]
```diff
@@ -54,37 +61,285 @@ static ColliderCylinderInit D_809CBA80 = {
...
+    } else if (curFrame < 12.0f) {
+        this->unk_1AE += 250.0f;
+    } else {
+        this->unk_1AE -= 250.0f;
+    }
+}
+
+void func_809CB218(EnBrob* this, GlobalContext* globalCtx) {
+    SkelAnime_Update(&this->skelAnime);
+    if (Animation_OnFrame(&this->skelAnime, 6.0f) != 0 || Animation_OnFrame(&this->skelAnime, 15.0f) != 0) {
```
Comment (petrie911) [inline src/overlays/actors/ovl_En_Brob/z_en_brob.c:197]:
```suggestion
    if (Animation_OnFrame(&this->skelAnime, 6.0f) || Animation_OnFrame(&this->skelAnime, 15.0f)) {
```

File: src/overlays/actors/ovl_En_Brob/z_en_brob.c [hunk 7]
```diff
@@ -54,37 +61,285 @@ static ColliderCylinderInit D_809CBA80 = {
...
+        Audio_PlayActorSound2(&this->dyna.actor, NA_SE_EN_BROB_WAVE);
+    }
+    if (this->timer != 0) {
+        this->timer--;
+    }
+    if (this->timer == 0 && 500.0f < this->dyna.actor.xzDistToPlayer) {
```
Comment (petrie911) [inline src/overlays/actors/ovl_En_Brob/z_en_brob.c:203]:
```suggestion
    if ((this->timer == 0) && (this->dyna.actor.xzDistToPlayer > 500.0f)) {
```

File: src/overlays/actors/ovl_En_Brob/z_en_brob.c [hunk 8]
```diff
@@ -54,37 +61,285 @@ static ColliderCylinderInit D_809CBA80 = {
...
+        func_809CAF88(this);
+    }
+}
+
+void func_809CB2B8(EnBrob* this, GlobalContext* globalCtx) {
+    if (SkelAnime_Update(&this->skelAnime) != 0) {
```
Comment (petrie911) [inline src/overlays/actors/ovl_En_Brob/z_en_brob.c:209]:
```suggestion
    if (SkelAnime_Update(&this->skelAnime)) {
```

File: src/overlays/actors/ovl_En_Brob/z_en_brob.c [hunk 9]
```diff
@@ -54,37 +61,285 @@ static ColliderCylinderInit D_809CBA80 = {
...
+    } else if (this->skelAnime.curFrame < 8.0f) {
+        this->unk_1AE -= 1250.0f;
+    }
+    this->dyna.actor.colorFilterTimer = 0x50;
+}
+
+void func_809CB354(EnBrob* this, GlobalContext* globalCtx) {
+    f32 curFrame;
+
+    if (SkelAnime_Update(&this->skelAnime) != 0) {
```
Comment (petrie911) [inline src/overlays/actors/ovl_En_Brob/z_en_brob.c:220]:
```suggestion
    if (SkelAnime_Update(&this->skelAnime)) {
```

File: src/overlays/actors/ovl_En_Brob/z_en_brob.c [hunk 10]
```diff
@@ -54,37 +61,285 @@ static ColliderCylinderInit D_809CBA80 = {
...
+        func_809CADDC(this, globalCtx);
+        return;
```
Comment (petrie911) [inline src/overlays/actors/ovl_En_Brob/z_en_brob.c:222]: also `else` here.

File: src/overlays/actors/ovl_En_Brob/z_en_brob.c [hunk 11]
```diff
@@ -54,37 +61,285 @@ static ColliderCylinderInit D_809CBA80 = {
...
+    f32 dist2;
+    s32 i;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Brob/func_809CAE44.s")
+    SkelAnime_Update(&this->skelAnime);
+    if (Animation_OnFrame(&this->skelAnime, 0) != 0) {
+        if (this->timer != 0) {
+            this->timer--;
+        }
+    }
```
Comment (petrie911) [inline src/overlays/actors/ovl_En_Brob/z_en_brob.c:246]:
```suggestion
    if (Animation_OnFrame(&this->skelAnime, 0) && (this->timer != 0)) {
        this->timer--;
    }
```

File: src/overlays/actors/ovl_En_Brob/z_en_brob.c [hunk 12]
```diff
@@ -54,37 +61,285 @@ static ColliderCylinderInit D_809CBA80 = {
...
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Brob/func_809CAEA0.s")
+    if (globalCtx->gameplayFrames % 2) {
+        dist1 = 0.0f;
+    } else {
+        dist1 = this->dyna.actor.scale.x * 5500.0f;
+    }
```
Comment (petrie911) [inline src/overlays/actors/ovl_En_Brob/z_en_brob.c:252]: possible ternary?

File: src/overlays/actors/ovl_En_Brob/z_en_brob.h [hunk 13]
```diff
@@ -6,9 +6,19 @@
 
 struct EnBrob;
 
+typedef void (*EnBrobUpdateFunc)(struct EnBrob* this, GlobalContext* globalCtx);
```
Comment (petrie911) [inline src/overlays/actors/ovl_En_Brob/z_en_brob.h:9]: This should be `ActionFunc`. `UpdateFunc` is used when the function is set in `Init` and stays fixed.

File: src/overlays/actors/ovl_En_Brob/z_en_brob.h [hunk 14]
```diff
@@ -6,9 +6,19 @@
...
-    /* 0x014C */ char unk_14C[0x174];
+    /* 0x0000 */ DynaPolyActor dyna;
+    /* 0x0164 */ SkelAnime skelAnime;
+    /* 0x01A8 */ EnBrobUpdateFunc updateFunc;
+    /* 0x01AC */ s16 timer;
+    /* 0x01AE */ s16 unk_1AE;
+    /* 0x01B0 */ Vec3s jointTable;
+    /* 0x01B6 */ char unk_1B6[0x36];
+    /* 0x01EC */ Vec3s morphTable;
+    /* 0x01F2 */ char unk_1F2[0x36];
```
Comment (petrie911) [inline src/overlays/actors/ovl_En_Brob/z_en_brob.h:20]: These are arrays.

File: src/overlays/actors/ovl_En_Brob/z_en_brob.c [hunk 15]
```diff
@@ -54,37 +61,285 @@ static ColliderCylinderInit D_809CBA80 = {
...
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Brob/func_809CB458.s")
+void EnBrob_PostLimbDraw(GlobalContext* globalCtx, s32 limbIndex, Gfx** dList, Vec3s* rot, void* thisx) {
+    EnBrob* this = THIS;
+    MtxF mtx;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Brob/EnBrob_Update.s")
+    Matrix_Get(&mtx);
+    if (limbIndex == 3) {
+        this->colliders[0].dim.pos.x = (s16)mtx.wx;
```
Comment (petrie911) [inline src/overlays/actors/ovl_En_Brob/z_en_brob.c:328]: these `s16` casts should be unneeded, as `dim.pos` is a Vec3s

File: src/overlays/actors/ovl_En_Brob/z_en_brob.c [hunk 16]
```diff
@@ -54,37 +61,285 @@ static ColliderCylinderInit D_809CBA80 = {
...
+            pos.x = this->dyna.actor.world.pos.x + dist1;
+            pos.z = this->dyna.actor.world.pos.z + dist2;
+        } else {
+            pos.x = this->dyna.actor.world.pos.x + dist2;
+            pos.z = this->dyna.actor.world.pos.z + dist1;
+            dist1 = -dist1;
+            dist2 = -dist2;
+        }
+        pos.y = (((Rand_ZeroOne() * 15000.0f) + 1000.0f) * this->dyna.actor.scale.y) + this->dyna.actor.world.pos.y;
+        EffectSsLightning_Spawn(globalCtx, &pos, &D_809CBAB4, &D_809CBAB8,
```
Comment (petrie911) [inline src/overlays/actors/ovl_En_Brob/z_en_brob.c:266]: could these colors be in-function static?

**Review by petrie911 — COMMENTED** [review]
Nice job, especially with whatever's going on with those colliders. Some suggestions.

**Comment by Emill** [comment]
Thanks for the quick review! All your changes work.

**Review by Roman971 — APPROVED** [review]

**Review by fig02 — APPROVED** [review]

**Pull request merged**